Add support for loading USDs in WCP and rendering them in GPUP

mwyrzykowski opened the pull request:
#### 7b221b942523779afedacf6f845e06507c929d09
<pre>
Add support for loading USDs in WCP and rendering them in GPUP
<a href="https://bugs.webkit.org/show_bug.cgi?id=298819">https://bugs.webkit.org/show_bug.cgi?id=298819</a>
<a href="https://rdar.apple.com/160530763">rdar://160530763</a>

Reviewed by Ada Chan.

Setup framework for sending geometry, transforms, textures -&gt; everything to render a USD,
from WCP to GPUP.

* Source/WebCore/Configurations/WebCore.xcconfig:
* Source/WebCore/Headers.cmake:
* Source/WebCore/Modules/Model/Implementation/DDMeshImpl.cpp: Added.
(WebCore::DDModel::DDMeshImpl::DDMeshImpl):
(WebCore::DDModel::DDMeshImpl::setLabelInternal):
(WebCore::DDModel::convertToBacking):
(WebCore::DDModel::DDMeshImpl::update):
* Source/WebCore/Modules/Model/Implementation/DDMeshImpl.h: Copied from Source/WebCore/Modules/WebGPU/Implementation/WebGPUCreateImpl.cpp.
* Source/WebCore/Modules/Model/Implementation/ModelConvertToBackingContext.cpp: Added.
* Source/WebCore/Modules/Model/Implementation/ModelConvertToBackingContext.h: Added.
* Source/WebCore/Modules/Model/Implementation/ModelDowncastConvertToBackingContext.cpp: Added.
(WebCore::DDModel::DowncastConvertToBackingContext::convertToBacking):
* Source/WebCore/Modules/Model/Implementation/ModelDowncastConvertToBackingContext.h: Added.
* Source/WebCore/Modules/Model/InternalAPI/DDMesh.h: Added.
(WebCore::DDModel::DDMesh::label const):
(WebCore::DDModel::DDMesh::setLabel):
* Source/WebCore/Modules/Model/InternalAPI/DDMeshDescriptor.h: Added.
* Source/WebCore/Modules/Model/InternalAPI/DDMeshPart.h: Added.
* Source/WebCore/Modules/Model/InternalAPI/DDReplaceVertices.h: Added.
* Source/WebCore/Modules/Model/InternalAPI/DDUpdateMeshDescriptor.h: Added.
* Source/WebCore/Modules/Model/InternalAPI/DDVertexAttributeFormat.h: Added.
* Source/WebCore/Modules/Model/InternalAPI/DDVertexLayout.h: Added.
* Source/WebCore/Modules/Model/InternalAPI/ModelObjectDescriptorBase.h: Added.
* Source/WebCore/Modules/WebGPU/Implementation/WebGPUCreateImpl.cpp:
(WebCore::WebGPU::create):
* Source/WebCore/Modules/WebGPU/Implementation/WebGPUImpl.cpp:
(WebCore::WebGPU::GPUImpl::GPUImpl):
(WebCore::WebGPU::convertToBacking):
(WebCore::WebGPU::GPUImpl::addMeshRequest):
* Source/WebCore/Modules/WebGPU/Implementation/WebGPUImpl.h:
* Source/WebCore/Modules/WebGPU/Implementation/WebGPUPtr.h:
* Source/WebCore/Modules/WebGPU/InternalAPI/WebGPU.h:
* Source/WebCore/Modules/mediasource/ManagedMediaSource.cpp:
* Source/WebCore/Modules/mediastream/RTCRtpSFrameTransformerCocoa.cpp:
* Source/WebCore/Sources.txt:
* Source/WebCore/SourcesCocoa.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/html/canvas/UsdModelLoader.swift: Added.
(USDModelLoader.timer):
(Host.state):
(Connected.hosts):
(state):
(meshes):
(primIdToMeshId):
(materials):
(primIdToMaterialId):
(materialIdToTextureIds):
(assetPathToTextureId):
(loadModel(from:)):
(addHost):
(updateHost(_:size:contentsScale:)):
(removeHost(_:)):
(removeMesh(_:)):
(setCameraDistance(_:)):
(setEnableModelRotation(_:)):
(meshCreated(_:)):
(meshUpdated(_:data:)):
(meshDestroyed(_:)):
(materialCreated(_:)):
(materialUpdated(_:data:)):
(materialDestroyed(_:)):
(textureCreated(_:)):
(textureUpdated(_:data:)):
(textureDestroyed(_:)):
(WebAddMeshRequest.indexCapacity):
(WebAddMeshRequest.indexType):
(WebAddMeshRequest.vertexBufferCount):
(WebAddMeshRequest.vertexCapacity):
(WebAddMeshRequest.vertexAttributes):
(WebAddMeshRequest.vertexLayouts):
(WebDDMeshPart.indexOffset):
(WebDDMeshPart.indexCount):
(WebDDMeshPart.topology):
(WebDDMeshPart.materialIndex):
(WebDDMeshPart.boundsMin):
(WebDDMeshPart.boundsMax):
(WebSetPart.partIndex):
(WebSetPart.part):
(WebSetRenderFlags.partIndex):
(WebSetRenderFlags.renderFlags):
(WebReplaceVertices.bufferIndex):
(WebReplaceVertices.buffer):
(WebUpdateMeshRequest.partCount):
(WebUpdateMeshRequest.parts):
(WebUpdateMeshRequest.renderFlags):
(WebUpdateMeshRequest.vertices):
(WebUpdateMeshRequest.indices):
(WebUpdateMeshRequest.transform):
(WebUpdateMeshRequest.instanceTransforms):
(WebUpdateMeshRequest.materialIds):
(WebUSDModelLoader.renderer):
(WebUSDModelLoader.modelAdded):
(WebUSDModelLoader.modelUpdated):
(WebUSDModelLoader.modelAddedCallback):
(WebUSDModelLoader.setCallbacks(_:modelUpdatedCallback:)):
(WebUSDModelLoader.loadModel(from:)):
(WebUSDModelLoader.toFloatArray(_:)):
(WebUSDModelLoader.results):
(WebUSDModelLoader.toWebSetParts(_:)):
(WebUSDModelLoader.toWebRenderFlags(_:)):
(WebUSDModelLoader.toWebReplaceVertices(_:)):
(WebUSDModelLoader.toWebVertexAttributes(_:)):
(WebUSDModelLoader.toWebVertexLayouts(_:)):
(WebUSDModelLoader.updateMesh(_:)):
(WebUSDModelLoader.addMesh(_:callback:)):
(WebUSDModelLoader.meshAdded(_:)):
(WebUSDModelLoader.updateHostedRenderer(_:size:contentsScale:)):
* Source/WebGPU/WebGPU.xcodeproj/project.pbxproj:
* Source/WebGPU/WebGPU/CommandEncoder.swift:
(WebGPU.beginRenderPass(_:)):
* Source/WebGPU/WebGPU/Instance.h:
* Source/WebGPU/WebGPU/Instance.mm:
(WebGPU::getDevices):
(WebGPU::m_isValid):
(WebGPU::Instance::requestAdapter):
(WebGPU::Instance::updateModel):
(WebGPU::convertPart):
(WebGPU::convertUint8s):
(WebGPU::convertDescriptor):
(WebGPU::Instance::updateMesh):
(wgpuDDMeshCreate):
* Source/WebGPU/WebGPU/Internal/Logging.h:
* Source/WebGPU/WebGPU/UsdModelRenderer.swift: Added.
(DDBridgeAddMeshRequest.indexCapacity):
(DDBridgeAddMeshRequest.indexType):
(DDBridgeAddMeshRequest.vertexBufferCount):
(DDBridgeAddMeshRequest.vertexCapacity):
(DDBridgeAddMeshRequest.vertexAttributes):
(DDBridgeAddMeshRequest.vertexLayouts):
(DDBridgeAddMeshRequest.makeDescriptor):
(DDMeshPart.indexOffset):
(DDMeshPart.indexCount):
(DDMeshPart.topology):
(DDMeshPart.materialIndex):
(DDMeshPart.boundsMin):
(DDMeshPart.boundsMax):
(DDBridgeSetPart.partIndex):
(DDBridgeSetPart.part):
(DDBridgeSetRenderFlags.partIndex):
(DDBridgeSetRenderFlags.renderFlags):
(DDBridgeReplaceVertices.bufferIndex):
(DDBridgeReplaceVertices.buffer):
(DDBridgeUpdateMesh.partCount):
(DDBridgeUpdateMesh.parts):
(DDBridgeUpdateMesh.renderFlags):
(DDBridgeUpdateMesh.vertices):
(DDBridgeUpdateMesh.indices):
(DDBridgeUpdateMesh.transform):
(DDBridgeUpdateMesh.instanceTransforms):
(DDBridgeUpdateMesh.materialIds):
(DDReceiver.device):
(DDReceiver.context):
(DDReceiver.scene):
(DDReceiver.meshes):
(DDReceiver.materials):
(DDReceiver.nodeDefinitionStore):
(DDReceiver.camera):
(DDReceiver.commandQueue):
(DDReceiver.captureManager):
(DDReceiver.setDevice(_:)):
(DDReceiver.makeCaptureDescriptor() -&gt; (MTLCaptureDescriptor:Foundation:)):
(DDReceiver.render(_:)):
(DDReceiver.addMesh(_:identifier:)):
(DDReceiver.updateMesh(_:identifier:)):
* Source/WebGPU/WebGPU/WebGPUExt.h:
* Source/WebKit/CMakeLists.txt:
* Source/WebKit/DerivedSources-input.xcfilelist:
* Source/WebKit/DerivedSources-output.xcfilelist:
* Source/WebKit/DerivedSources.make:
* Source/WebKit/GPUProcess/graphics/Model/ModelObjectHeap.cpp: Added.
(WebKit::DDModel::ObjectHeap::ObjectHeap):
(WebKit::DDModel::ObjectHeap::addObject):
(WebKit::DDModel::ObjectHeap::removeObject):
(WebKit::DDModel::ObjectHeap::clear):
(WebKit::DDModel::ObjectHeap::convertDDMeshFromBacking):
(WebKit::DDModel::ObjectHeap::objectExistsAndValid const):
* Source/WebKit/GPUProcess/graphics/Model/ModelObjectHeap.h: Added.
* Source/WebKit/GPUProcess/graphics/Model/RemoteDDMesh.cpp: Copied from Source/WebKit/GPUProcess/graphics/WebGPU/RemoteBindGroup.cpp.
(WebKit::RemoteDDMesh::RemoteDDMesh):
(WebKit::RemoteDDMesh::connection const):
(WebKit::RemoteDDMesh::stopListeningForIPC):
(WebKit::RemoteDDMesh::destruct):
(WebKit::RemoteDDMesh::setLabel):
(WebKit::RemoteDDMesh::protectedBacking):
(WebKit::RemoteDDMesh::update):
* Source/WebKit/GPUProcess/graphics/Model/RemoteDDMesh.h: Added.
* Source/WebKit/GPUProcess/graphics/Model/RemoteDDMesh.messages.in: Copied from Source/WebKit/GPUProcess/graphics/WebGPU/RemoteGPU.messages.in.
* Source/WebKit/GPUProcess/graphics/WebGPU/RemoteBindGroup.cpp:
* Source/WebKit/GPUProcess/graphics/WebGPU/RemoteGPU.cpp:
(WebKit::RemoteGPU::RemoteGPU):
(WebKit::RemoteGPU::workQueueUninitialize):
(WebKit::RemoteGPU::addMeshRequest):
* Source/WebKit/GPUProcess/graphics/WebGPU/RemoteGPU.h:
* Source/WebKit/GPUProcess/graphics/WebGPU/RemoteGPU.messages.in:
* Source/WebKit/GPUProcess/webrtc/RemoteAudioMediaStreamTrackRendererInternalUnitManager.cpp:
(WebKit::RemoteAudioMediaStreamTrackRendererInternalUnitManagerUnit::RemoteAudioMediaStreamTrackRendererInternalUnitManagerUnit):
* Source/WebKit/Scripts/webkit/messages.py:
(atomic_object_identifier):
(serialized_identifiers):
(headers_for_type):
* Source/WebKit/Shared/Model/DDMeshDescriptor.cpp: Added.
(WebKit::DDModel::ConvertToBackingContext::convertToBacking):
(WebKit::DDModel::convertVectorToBacking):
(WebKit::DDModel::ConvertFromBackingContext::convertFromBacking):
(WebKit::DDModel::convertVectorFromBacking):
* Source/WebKit/Shared/Model/DDMeshDescriptor.h: Added.
* Source/WebKit/Shared/Model/DDMeshDescriptor.serialization.in: Added.
* Source/WebKit/Shared/Model/DDMeshPart.cpp: Added.
(WebKit::DDModel::ConvertToBackingContext::convertToBacking):
(WebKit::DDModel::ConvertFromBackingContext::convertFromBacking):
* Source/WebKit/Shared/Model/DDMeshPart.h: Added.
* Source/WebKit/Shared/Model/DDMeshPart.serialization.in: Added.
* Source/WebKit/Shared/Model/DDModelIdentifier.h: Added.
* Source/WebKit/Shared/Model/DDReplaceVertices.cpp: Added.
(WebKit::DDModel::ConvertToBackingContext::convertToBacking):
(WebKit::DDModel::ConvertFromBackingContext::convertFromBacking):
* Source/WebKit/Shared/Model/DDReplaceVertices.h: Added.
* Source/WebKit/Shared/Model/DDReplaceVertices.serialization.in: Added.
* Source/WebKit/Shared/Model/DDUpdateMeshDescriptor.cpp: Added.
(WebKit::DDModel::ConvertToBackingContext::convertToBacking):
(WebKit::DDModel::convertVectorToBacking):
(WebKit::DDModel::ConvertFromBackingContext::convertFromBacking):
(WebKit::DDModel::convertVectorFromBacking):
* Source/WebKit/Shared/Model/DDUpdateMeshDescriptor.h: Added.
* Source/WebKit/Shared/Model/DDUpdateMeshDescriptor.serialization.in: Added.
* Source/WebKit/Shared/Model/DDVertexAttributeFormat.h: Added.
* Source/WebKit/Shared/Model/DDVertexAttributeFormat.serialization.in: Added.
* Source/WebKit/Shared/Model/DDVertexLayout.h: Added.
* Source/WebKit/Shared/Model/DDVertexLayout.serialization.in: Added.
* Source/WebKit/Shared/Model/ModelConvertFromBackingContext.h: Added.
* Source/WebKit/Shared/Model/ModelConvertToBackingContext.h: Added.
* Source/WebKit/Shared/Model/ModelObjectDescriptorBase.cpp: Added.
(WebKit::DDModel::ConvertToBackingContext::convertToBacking):
(WebKit::DDModel::ConvertFromBackingContext::convertFromBacking):
* Source/WebKit/Shared/Model/ModelObjectDescriptorBase.h: Added.
* Source/WebKit/Shared/Model/ModelObjectDescriptorBase.serialization.in: Added.
* Source/WebKit/Shared/WTFArgumentCoders.serialization.in:
* Source/WebKit/Sources.txt:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/GPU/graphics/Model/ModelDowncastConvertToBackingContext.cpp: Added.
(WebKit::DDModel::DowncastConvertToBackingContext::convertToBacking):
* Source/WebKit/WebProcess/GPU/graphics/Model/ModelDowncastConvertToBackingContext.h: Added.
* Source/WebKit/WebProcess/GPU/graphics/Model/RemoteDDMeshProxy.cpp: Added.
(WebKit::DDModel::RemoteDDMeshProxy::RemoteDDMeshProxy):
(WebKit::DDModel::RemoteDDMeshProxy::~RemoteDDMeshProxy):
(WebKit::DDModel::RemoteDDMeshProxy::update):
(WebKit::DDModel::RemoteDDMeshProxy::setLabelInternal):
* Source/WebKit/WebProcess/GPU/graphics/Model/RemoteDDMeshProxy.h: Added.
* Source/WebKit/WebProcess/GPU/graphics/WebGPU/RemoteGPUProxy.cpp:
(WebKit::RemoteGPUProxy::create):
(WebKit::RemoteGPUProxy::RemoteGPUProxy):
(WebKit::RemoteGPUProxy::addMeshRequest):
* Source/WebKit/WebProcess/GPU/graphics/WebGPU/RemoteGPUProxy.h:
* Source/WebKit/WebProcess/GPU/media/RemoteAudioDestinationProxy.cpp:
(WebKit::RemoteAudioDestinationProxy::outputLatency const):
(WebKit::RemoteAudioDestinationProxy::renderAudio):
* Source/WebKit/WebProcess/Network/webrtc/LibWebRTCNetworkManager.cpp:
* Source/WebKit/WebProcess/Storage/WebSWContextManagerConnection.cpp:
* Source/WebKit/WebProcess/WebCoreSupport/WebChromeClient.cpp:
(WebKit::WebChromeClient::createGPUForWebGPU const):
* Source/WebKit/WebProcess/WebCoreSupport/WebPermissionController.cpp:
(WebKit::WebPermissionController::query):
* Source/WebKit/WebProcess/WebCoreSupport/WebWorkerClient.cpp:
(WebKit::GPUProcessWebWorkerClient::createGPUForWebGPU const):

Canonical link: <a href="https://commits.webkit.org/300626@main">https://commits.webkit.org/300626@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/fa3ac5bf4837404a954b6cca5e37a23083870aa1

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/123282 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/42996 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/33693 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/129990 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/75395 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/13403a96-1595-40f5-ab3f-e882e9bbe1cf) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/125159 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/43720 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/51591 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/93710 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/62166 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/eb8c9782-f951-4d07-b735-2c9f4ff336e0) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/126235 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/34837 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/110307 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/74338 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/9fb3c0dc-d633-40f6-9b38-19c9f7a5547f) 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/122638 "Passed tests") | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/33802 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/28466 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/73508 "Built successfully") | | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/115441 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/104551 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/28692 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/132708 "Built successfully") | | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/121813 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/50232 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/38225 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/102200 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/50608 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/106531 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/102055 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/25947 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/47408 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/25630 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/47006 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/50087 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/55848 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/152168 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/49558 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/38885 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/52908 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/51236 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

mwyrzykowski commented on Source/WebCore/Modules/mediasource/ManagedMediaSource.cpp:
```diff
@@ -32,6 +32,7 @@
 #include "EventNames.h"
 #include "ExceptionOr.h"
 #include "MediaSourcePrivate.h"
+#include "ScriptExecutionContext.h"
```
unified build fix

mwyrzykowski commented on Source/WebCore/Modules/mediastream/RTCRtpSFrameTransformerCocoa.cpp:
```diff
@@ -29,6 +29,7 @@
 #if ENABLE(WEB_RTC)
 
 #include "CryptoUtilitiesCocoa.h"
+#include "ExceptionOr.h"
```
unified build fix

mwyrzykowski commented on Source/WebKit/GPUProcess/graphics/WebGPU/RemoteGPU.h:
```diff
@@ -121,6 +126,7 @@ class RemoteGPU final : public CanMakeWeakPtr<RemoteGPU>, public IPC::StreamServ
 
 
     void requestAdapter(const WebGPU::RequestAdapterOptions&, WebGPUIdentifier, CompletionHandler<void(std::optional<RemoteGPURequestAdapterResponse>&&)>&&);
+    void addMeshRequest(const DDModel::DDMeshDescriptor&, WebKit::DDModelIdentifier);
```
We intentionally use RemoteGPU for two reasons:

(1) rendering the models requires access to a MTLDevice
(2) we want the rendering to occur on the same thread as other GPU tasks so <model> can eventually take advantage of WGSL shaders

rr-codes commented on Source/WebGPU/WebGPU/UsdModelRenderer.swift:
```diff
@@ -0,0 +1,437 @@
+import ShaderGraph
+#endif
+
+@available(macOS 26.0, iOS 26.0, *)
```
availability annotations are not needed since none of these types are public

pxlcoder reviewed (commented):
This PR is quite large and difficult to review, could it be split up into multiple PRs?

I have not read through fully enough to know if this is viable, but how about:
1. Introduce new classes / "API" surface
2. Introduce IPC / MessageReceiver changes as stub implementations
3. Implement Swift changes
4. Fill in IPC changes

pxlcoder commented on Source/WebGPU/WebGPU/Instance.mm:
```diff
@@ -203,6 +219,111 @@
+    current.transform = fs[0];
+
+    for (uint32_t i = 1; i < count; ++i) {
+        DDBridgeChainedFloat4x4* next = [[DDBridgeChainedFloat4x4 alloc] init];
```
Many instances throughout the PR:
```suggestion
        DDBridgeChainedFloat4x4 *next = [[DDBridgeChainedFloat4x4 alloc] init];
```

mwyrzykowski replied:
We really need a way for check-webkit-style to enforce otherwise the coding style should allow either style

mwyrzykowski replied:
Thank you :) I tried to identify and correct all of these.

rr-codes commented on Source/WebCore/Modules/Model/Implementation/DDMeshImpl.cpp:
```diff
@@ -0,0 +1,103 @@
+/*
+ * Copyright (C) 2021-2023 Apple Inc. All rights reserved.
```
should this copyright (and the others) have 2025?

rr-codes commented on Source/WebCore/Modules/Model/Implementation/DDMeshImpl.h:
```diff
@@ -0,0 +1,78 @@
+
+#if HAVE(WEBGPU_IMPLEMENTATION)
+
+#include "DDMesh.h"
```
All private headers must be included via `<WebCore/whatever.h>` to comply with modules

rr-codes commented on Source/WebCore/Modules/Model/InternalAPI/DDFloat3.h:
```diff
@@ -0,0 +1,62 @@
+namespace WebCore::DDModel {
+
+struct DDFloat3 {
+    union {
```
should this be WTF::Variant instead of a C union?

mwyrzykowski replied:
No because we want the memory to be shared. We don't want it to be a variant

rr-codes commented on Source/WebCore/Modules/Model/InternalAPI/DDMeshDescriptor.h:
```diff
@@ -0,0 +1,43 @@
+namespace WebCore::DDModel {
+
+struct DDMeshDescriptor : public ObjectDescriptorBase {
+    int32_t indexCapacity;
```
should these have default values of `0` to avoid undefined behavior if they are not all initialized?

rr-codes commented on Source/WebCore/Modules/Model/ModelDDInlineConverters.h:
```diff
@@ -0,0 +1,167 @@
+    };
+}
+
+static Vector<DDModel::DDVertexAttributeFormat> toCppVertexAttributes(NSArray<WebDDVertexAttributeFormat *> *arr)
```
I think we prefer to use more descriptive parameter names, or at the very least the full word instead of abbreviations like “arr"

rr-codes commented on Source/WebCore/Modules/Model/ModelDDInlineConverters.h:
```diff
@@ -0,0 +1,167 @@
+#include <WebCore/DDUpdateMeshDescriptor.h>
+
+namespace WebCore {
+static DDModel::DDVertexAttributeFormat toCpp(WebDDVertexAttributeFormat *p)
```
What does `p` mean?

mwyrzykowski replied:
`pointer` but I will change the name to `format`

rr-codes commented on Source/WebCore/Modules/Model/ModelDDInlineConverters.h:
```diff
@@ -0,0 +1,167 @@
+static Vector<KeyValuePair<int32_t, uint64_t>> toCpp(NSArray<WebSetRenderFlags*> *renderFlags)
+{
+    Vector<KeyValuePair<int32_t, uint64_t>> result;
+    for (WebSetRenderFlags* f in renderFlags)
```
maybe `flag` instead of `f`?

rr-codes commented on Source/WebCore/Modules/Model/ModelDDInlineConverters.h:
```diff
@@ -0,0 +1,167 @@
+{
+    return DDModel::DDMeshPart {
+        { "mesh part"_s, },
+        (uint32_t)part.indexOffset,
```
I think we prefer to use static_cast instead of C-style casts

rr-codes commented on Source/WebCore/Modules/Model/ModelDDInlineConverters.h:
```diff
@@ -0,0 +1,167 @@
+    return result;
+}
+
+static Vector<String> toCpp(NSArray<NSUUID*> *arr)
```
```suggestion
static Vector<String> toCpp(NSArray<NSUUID *> *arr)
```

rr-codes commented on Source/WebCore/Modules/Model/ModelDDInlineConverters.h:
```diff
@@ -0,0 +1,167 @@
+    };
+}
+
+static Vector<DDModel::DDReplaceVertices> toCpp(NSArray<WebReplaceVertices*> *arr)
```
```suggestion
static Vector<DDModel::DDReplaceVertices> toCpp(NSArray<WebReplaceVertices *> *arr)
```

rr-codes commented on Source/WebCore/Modules/Model/ModelDDTypes.h:
```diff
@@ -0,0 +1,153 @@
+@property (nonatomic) int32_t indexType;
+@property (nonatomic) int32_t vertexBufferCount;
+@property (nonatomic) int32_t vertexCapacity;
+@property (nonatomic, strong, nullable) NSArray<WebDDVertexAttributeFormat*> *vertexAttributes;
```
strong is the default, so it’s not necessary to explicitly specify it

mwyrzykowski replied:
Swift complained when I left out `strong`

mwyrzykowski replied:
I can't reproduce the error I observed earlier so I remove it. I did update my toolchain recently, unsure if that was related

rr-codes commented on Source/WebCore/Modules/Model/ModelDDTypes.h:
```diff
@@ -0,0 +1,153 @@
+- (instancetype)init NS_UNAVAILABLE;
+- (instancetype)initWithValue:(int32_t)value NS_DESIGNATED_INITIALIZER;
+
+- (void)loadModelFrom:(NSURL*)url;
```
```suggestion
- (void)loadModelFrom:(NSURL *)url;
```

rr-codes commented on Source/WebCore/WebCore.xcodeproj/project.pbxproj:
```diff
@@ -389,10 +389,25 @@
 		0DA042502E417B1500683D92 /* GPUTextureFormat.h in Headers */ = {isa = PBXBuildFile; fileRef = 1C66A7C42A4F78580086DB6F /* GPUTextureFormat.h */; settings = {ATTRIBUTES = (Private, ); }; };
 		0DA042512E417B3500683D92 /* GPUTextureUsage.h in Headers */ = {isa = PBXBuildFile; fileRef = 1C66A6FB2A4F78170086DB6F /* GPUTextureUsage.h */; settings = {ATTRIBUTES = (Private, ); }; };
 		0DA042522E417B4800683D92 /* GPUIntegralTypes.h in Headers */ = {isa = PBXBuildFile; fileRef = 1C66A7DA2A4F785E0086DB6F /* GPUIntegralTypes.h */; settings = {ATTRIBUTES = (Private, ); }; };
+		0DACD8592E79D8C100CDA0DC /* DDModel.serialization.in in Headers */ = {isa = PBXBuildFile; fileRef = 0DACD8582E79D8C100CDA0DC /* DDModel.serialization.in */; settings = {ATTRIBUTES = (Private, ); }; };
```
can any of these be Project instead of Private? (I assume not but just want to confirm)

mwyrzykowski replied:
I think the ones which are Private are all needed in WebKit.framework

rr-codes commented on Source/WebKit/GPUProcess/graphics/Model/ModelObjectHeap.h:
```diff
@@ -0,0 +1,86 @@
+private:
+    ObjectHeap();
+
+    using Object = Variant<
```
what’s the difference between this and using std::optional?

mwyrzykowski replied:
Only we will need to add more types here so I copied the existing pattern in WebGPUObjectHeap

rr-codes commented on Source/WebGPU/WebGPU/UsdModelRenderer.swift:
```diff
@@ -0,0 +1,428 @@
+import Foundation
```
all of these should be `internal import` if possible

mwyrzykowski replied:
I can't make these internal import without encountering:
```
error: ambiguous implicit access level for import of 'Metal'; it is imported as 'internal' elsewhere
``

mwyrzykowski replied:
I don't think it matters because this is inside a private framework

rr-codes commented on Source/WebGPU/WebGPU/UsdModelRenderer.swift:
```diff
@@ -0,0 +1,428 @@
+        self.format = format
+        self.layoutIndex = layoutIndex
+        self.offset = offset
+        super.init()
```
I _think_ the `super.init()` isn’t needed? I’m not sure, but if not then it should be removed

mwyrzykowski replied:
yes it looks like I can remove the super.init calls

mwyrzykowski replied:
except in WebUSDModelLoader

mwyrzykowski replied:
due to
```
Volumes/Data/OpenSource/Source/WebCore/html/canvas/UsdModelLoader.swift:610:25: error: 'self' used before 'super.init' call
        self.renderer = USDModelLoader(objcInstance: self)
```

rr-codes commented on Source/WebGPU/WebGPU/UsdModelRenderer.swift:
```diff
@@ -0,0 +1,428 @@
+@objc
+@implementation
+extension DDBridgeAddMeshRequest {
+    var indexCapacity: Int32
```
are these ever mutated? If not, they can be `let`

mwyrzykowski replied:
they are not, I just need to make an initializer

rr-codes commented on Source/WebGPU/WebGPU/UsdModelRenderer.swift:
```diff
@@ -0,0 +1,428 @@
+    var vertexAttributes: [DDBridgeVertexAttributeFormat]?
+    var vertexLayouts: [DDBridgeVertexLayout]?
+
+    init(value: Int32) {
```
where is `value` used here?

mwyrzykowski replied:
it's not used, but upon removing it I saw:
```
/Volumes/Data/OpenSource/Source/WebGPU/WebGPU/UsdModelRenderer.swift:61:5: error: overriding declaration requires an 'override' keyword
    init() {
/Volumes/Data/XcodeB.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX26.1.Internal.sdk/usr/include/objc/NSObject.h:66:1: note: overridden declaration is here
- (instancetype)init
^
```

because I had forgotten `override` on the Swift method. 

Adding `override` it seems to be fine with that change applied :)

rr-codes commented on Source/WebGPU/WebGPU/UsdModelRenderer.swift:
```diff
@@ -0,0 +1,428 @@
+        self.vertexCapacity = 0
+        super.init()
+    }
+    #if canImport(DirectDrawBackend)
```
newline before this line please

rr-codes commented on Source/WebGPU/WebGPU/UsdModelRenderer.swift:
```diff
@@ -0,0 +1,428 @@
+        super.init()
+    }
+    #if canImport(DirectDrawBackend)
+    fileprivate func makeDescriptor() -> DirectResource.MeshDescriptor {
```
```suggestion
    fileprivate var descriptor: DirectResource.MeshDescriptor {
```

rule of thumb: generally only use functions when I/O is being performed or side effects happen, otherwise properties are usually preferred

rr-codes commented on Source/WebGPU/WebGPU/UsdModelRenderer.swift:
```diff
@@ -0,0 +1,428 @@
+        descriptor.indexType = MTLIndexType(rawValue: UInt(indexType)) ?? .uint16
+        descriptor.vertexBufferCount = DRInteger(vertexBufferCount)
+        descriptor.vertexCapacity = DRInteger(vertexCapacity)
+        if let vertexAttributes = self.vertexAttributes {
```
```suggestion
        if let vertexAttributes {
```

rr-codes commented on Source/WebGPU/WebGPU/UsdModelRenderer.swift:
```diff
@@ -0,0 +1,428 @@
+                )
+            }
+        }
+        if let vertexLayouts = self.vertexLayouts {
```
ditto

rr-codes commented on Source/WebGPU/WebGPU/UsdModelRenderer.swift:
```diff
@@ -0,0 +1,428 @@
+    fileprivate var meshes: [UUID: DirectDrawBackend.MeshInstance] = [:]
+    @nonobjc
+    fileprivate var materials: [UUID: DirectDrawBackend.MaterialInstance] = [:]
+    // var textures: [UUID:(DirectDrawBackend.TextureResource, String)] = [:]
```
please remove the commented out lines

rr-codes commented on Source/WebGPU/WebGPU/UsdModelRenderer.swift:
```diff
@@ -0,0 +1,428 @@
+    override init() {
+        super.init()
+    }
+    @objc(setDeviceWithDevice:)
```
newline before this line

rr-codes commented on Source/WebGPU/WebGPU/UsdModelRenderer.swift:
```diff
@@ -0,0 +1,428 @@
+        self.device = device
+
+        #if canImport(DirectDrawBackend)
+        self.context = DirectDrawBackend.Context.createContext(device: device)
```
prefer to not use `self.` explicitly

rr-codes commented on Source/WebGPU/WebGPU/UsdModelRenderer.swift:
```diff
@@ -0,0 +1,428 @@
+    }
+}
+
+class Helper {
```
this should probably be something more like

```
extension MTLDevice {
    fileprivate var captureDescriptor: (MTLCaptureDescriptor, Foundation.URL) { 
    
```

instead of a random Helper type

mwyrzykowski replied:
This seems wrong since the capture descriptor is not a property of MTLDevice, it shouldn't be an extension

rr-codes replied:
another alternative is making an extension on `MTLCaptureDescriptor` that’s like

```
extension MTLCaptureDescriptor {
    fileprivate init(from device: MTLDevice)
}
```

and then later:

```
let descriptor = MTLCaptureDescriptor(from: whatever)
let outputURL = descriptor.outputURL
```

(Both of these are extremely common idioms/paradigms in Swift)

mwyrzykowski replied:
```
fileprivate convenience init(from device: MTLDevice) {
```
sounds like a good idea :)

rr-codes commented on Source/WebGPU/WebGPU/UsdModelRenderer.swift:
```diff
@@ -0,0 +1,428 @@
+class Helper {
+    static fileprivate func makeCaptureDescriptor(device: MTLDevice?) -> (MTLCaptureDescriptor, Foundation.URL) {
+        let now = Date()
+        let dateFormatter = DateFormatter()
```
Please use the modern Swift APIs for date formatting https://developer.apple.com/documentation/foundation/formatstyle

mwyrzykowski replied:
This code can just be removed

rr-codes commented on Source/WebGPU/WebGPU/UsdModelRenderer.swift:
```diff
@@ -0,0 +1,428 @@
+
+        self.context?.update(deltaTime: 1.0 / 60.0)
+        self.camera?.renderTarget = .texture(texture)
+        guard let scene = self.scene else {
```
```suggestion
        guard let scene else {
```

rr-codes commented on Source/WebGPU/WebGPU/UsdModelRenderer.swift:
```diff
@@ -0,0 +1,428 @@
+
+        print("Captured gpu trace at \(outputURL)")
+        captureManager?.stopCapture()
+        #endif
```
newline before this line

mwyrzykowski replied:
will fix but linter should catch this since we can't enforce non-linter rules consistently

rr-codes replied:
I mean it’s a bit like NSLog, which afaik doesn’t get checked by the style checker in Obj-C land even though it too should be avoided? I could be wrong though

rr-codes replied:
oh oops wrong comment lol

rr-codes commented on Source/WebGPU/WebGPU/UsdModelRenderer.swift:
```diff
@@ -0,0 +1,428 @@
+    @objc(addMesh:identifier:)
+    func addMesh(_ descriptor: DDBridgeAddMeshRequest, identifier: UUID) -> Bool {
+        #if canImport(DirectDrawBackend)
+        let meshDescriptor = descriptor.makeDescriptor()
```
descriptor isnt really a descriptor isnt it just a request which has a descriptor on it?

mwyrzykowski replied:
`descriptor` is our descriptor and `meshDescriptor` is RealityKit's

rr-codes commented on Source/WebGPU/WebGPU/UsdModelRenderer.swift:
```diff
@@ -0,0 +1,428 @@
+        #if canImport(DirectDrawBackend)
+        let meshDescriptor = descriptor.makeDescriptor()
+        do {
+            guard let context = self.context else {
```
most of this can be moved outside the do/catch

rr-codes commented on Source/WebGPU/WebGPU/UsdModelRenderer.swift:
```diff
@@ -0,0 +1,428 @@
+            scene.meshes.append(meshInstance)
+            return true
+        } catch {
+            print("unexpected error")
```
please use the standard Swift Logger facilitates to print/log stuff

rr-codes commented on Source/WebGPU/WebGPU/UsdModelRenderer.swift:
```diff
@@ -0,0 +1,428 @@
+            // }
+            var instTransforms: [simd_float4x4] = []
+            var chainedTransform = request.instanceTransforms
+            while chainedTransform != nil {
```
```suggestion
            while let chainedTransform {
```

mwyrzykowski replied:
not sure how to apply this because I assign to chainedTransform in the loop `chainedTransform = chainedTransform?.next` and observe:

```
Cannot assign to value: 'chainedTransform' is a 'let' constant
```

rr-codes replied:
oh sorry I meant while var chainedTransform

mwyrzykowski replied:
```
Optional type 'DDBridgeChainedFloat4x4?' cannot be used as a boolean; test for '!= nil' instead
```

rr-codes replied:
hmmm, and that's with `while var chainedTransform { ...`? This is a valid Swift construct, so I'm not sure why that error is happening

rr-codes commented on Source/WebGPU/WebGPU/WebGPUExt.h:
```diff
@@ -64,6 +68,55 @@ typedef enum WGPUSTypeExtended {
+    const simd_float4x4* instanceTransforms4x4;
+    uint32_t instanceTransforms4x4Count;
+    Vector<String> materialIds;
+
```
extra newline

rr-codes commented on Source/WebGPU/WebGPU/UsdModelRenderer.swift:
```diff
@@ -0,0 +1,439 @@
+@objc
+@implementation
+extension DDBridgeMeshPart {
+    var indexOffset: UInt
```
can all these be `let`?

rr-codes commented on Source/WebGPU/WebGPU/UsdModelRenderer.swift:
```diff
@@ -0,0 +1,439 @@
+@objc
+@implementation
+extension DDBridgeSetPart {
+    var partIndex: Int
```
ditto

rr-codes replied:
ditto

rr-codes commented on Source/WebGPU/WebGPU/UsdModelRenderer.swift:
```diff
@@ -0,0 +1,439 @@
+@objc
+@implementation
+extension DDBridgeSetRenderFlags {
+    var partIndex: Int
```
ditto

rr-codes commented on Source/WebGPU/WebGPU/UsdModelRenderer.swift:
```diff
@@ -0,0 +1,439 @@
+}
+
+class Helper {
+    static fileprivate func isNonZero(value: Float) -> Bool {
```
these should all be extension properties on Float/simd_float/etc

mwyrzykowski replied:
commented below but not sure if this is possible

rr-codes commented on Source/WebGPU/WebGPU/UsdModelRenderer.swift:
```diff
@@ -0,0 +1,439 @@
+        isNonZero(_: matrix.columns.0) || isNonZero(_: matrix.columns.1) || isNonZero(_: matrix.columns.2) || isNonZero(_: matrix.columns.3)
+    }
+
+    fileprivate static let logger: Logger = Logger(subsystem: "com.apple.WebKit", category: "model [GPU]")
```
this is conventionally done via

```
extension Logger {
    static let modelGPU = Logger(subsystem: "com.apple.WebKit", category: "model [GPU]”)
}
```

rr-codes commented on Source/WebCore/html/canvas/UsdModelLoader.swift:
```diff
@@ -0,0 +1,678 @@
+#if canImport(USDDirectDrawIO) && canImport(WebCore_Internal)
```
needs copyright

rr-codes commented on Source/WebCore/html/canvas/UsdModelLoader.swift:
```diff
@@ -0,0 +1,678 @@
+internal import USDDirectDrawIO
+internal import os
+
+class USDModelLoader: DDRenderDelegate {
```
all Swift classes should always be `final`

rr-codes replied:
final class

rr-codes commented on Source/WebCore/html/canvas/UsdModelLoader.swift:
```diff
@@ -0,0 +1,678 @@
+    class Mesh: Identifiable {
+        let id: UUID
+        let descriptor: DDMeshDescriptor
+        enum State {
```
nested types should precede variables/properties

rr-codes commented on Source/WebCore/html/canvas/UsdModelLoader.swift:
```diff
@@ -0,0 +1,678 @@
+
+    @Observable
+    class Material: Identifiable {
+        let id: UUID
```
Conventionally, when there are multiple different types all with the same id type, the idiom is doing something like this:

```
final class Material: Identifiable {
    struct ID: Hashable {
      private let value: UUID
    }
    
    let id: ID
    
}
```

so that it then becomes a unique opaque type that can’t be confused with other ids

mwyrzykowski replied:
I think this should remain UUID for now as that is the interface from RE and we need the same UUID on both sides.

rr-codes commented on Source/WebCore/html/canvas/UsdModelLoader.swift:
```diff
@@ -0,0 +1,678 @@
+        }
+    }
+    private(set) var materials: [Material] = []
+    private(set) var primIdToMaterialId: [String: UUID] = [:]
```
this should be `String: Material.ID` (and ditto for the rest)

mwyrzykowski replied:
I think this should remain UUID for now as that is the interface from RE and we need the same UUID on both sides.

rr-codes commented on Source/WebCore/html/canvas/UsdModelLoader.swift:
```diff
@@ -0,0 +1,678 @@
+    }
+
+    func loadModel(from url: URL) {
+        self.renderer.loadModel(from: url)
```
no `self.`

rr-codes commented on Source/WebCore/html/canvas/UsdModelLoader.swift:
```diff
@@ -0,0 +1,678 @@
+            self.state = state
+        }
+    }
+    private(set) var hosts: [Host] = []
```
should this be a Set<Host> or a `[Host.ID : Host]` instead?

rr-codes commented on Source/WebCore/html/canvas/UsdModelLoader.swift:
```diff
@@ -0,0 +1,678 @@
+}
+
+class Converter {
+    static func toWebSetParts(parts: [UpdateMeshRequest.SetPart]?) -> [WebSetPart]? {
```
using “Converter” types is an anti-pattern in Swift; instead, the convention is to define extension initializers on the type that accept the type that will be converted

mwyrzykowski replied:
Programming patterns are generally language agnostic.

mwyrzykowski replied:
I am not sure how extension initializer would help when these are being initialized from an Array, wouldn't we still need some function to populate the array?

rr-codes commented on Source/WebCore/html/canvas/UsdModelLoader.swift:
```diff
@@ -0,0 +1,678 @@
+class Converter {
+    static func toWebSetParts(parts: [UpdateMeshRequest.SetPart]?) -> [WebSetPart]? {
+        var results: [WebSetPart] = []
+        if let p = parts {
```
Using guard let here makes more sense

rr-codes commented on Source/WebGPU/WebGPU/UsdModelRenderer.swift:
```diff
@@ -0,0 +1,439 @@
+
+        captureObject = device
+        destination = .gpuTraceDocument
+        var url = URL.temporaryDirectory.appending(path: "capture.gputrace").standardizedFileURL
```
let

rr-codes commented on Source/WebGPU/WebGPU/UsdModelRenderer.swift:
```diff
@@ -0,0 +1,439 @@
+import Foundation
```
needs copyright

rr-codes commented on Source/WebGPU/WebGPU/UsdModelRenderer.swift:
```diff
@@ -0,0 +1,439 @@
+import Metal
+import QuartzCore
+import QuartzCore_Private
+import WebGPU_Internal
```
this one if nothing else should at least be `internal import` if possible

mwyrzykowski replied:
This causes a compilation error unless we want to change all the other WebGPU swift files to internal import as well

rr-codes replied:
ah true

rr-codes commented on Source/WebGPU/WebGPU/UsdModelRenderer.swift:
```diff
@@ -0,0 +1,439 @@
+@implementation
+extension DDBridgeAddMeshRequest {
+    var indexCapacity: Int32
+    var indexType: Int32
```
these all should be let

rr-codes commented on Source/WebGPU/WebGPU/UsdModelRenderer.swift:
```diff
@@ -0,0 +1,439 @@
+@implementation
+extension DDBridgeSetRenderFlags {
+    var partIndex: Int
+
```
no newline; and ditto

mwyrzykowski replied:
Will change but if check-webkit-style does not catch this then we will continue to have inconsistencies of this nature.

rr-codes replied:
I think it’s one of those things where we need to just get into the habit of using let by default; the style checker doesn’t know if this variable is being mutated later on or not. It’s like annotating an obj c property with readonly or not

rr-codes commented on Source/WebGPU/WebGPU/UsdModelRenderer.swift:
```diff
@@ -0,0 +1,439 @@
+@objc
+@implementation
+extension DDBridgeChainedFloat4x4 {
+    var transform: simd_float4x4
```
ditto

rr-codes commented on Source/WebGPU/WebGPU/UsdModelRenderer.swift:
```diff
@@ -0,0 +1,439 @@
+@objc
+@implementation
+extension DDBridgeUpdateMesh {
+    var partCount: Int
```
ditto

rr-codes commented on Source/WebGPU/WebGPU/UsdModelRenderer.swift:
```diff
@@ -0,0 +1,439 @@
+    var indices: Data?
+    var transform: simd_float4x4
+    var instanceTransforms: DDBridgeChainedFloat4x4? // array of float4x4
+    var materialIds: [UUID]?
```
this should be a typed identifier instead of a plain UUID

mwyrzykowski replied:
I don't think this suggestion is ideal as these are the types we get from RE.

So we would just be further boxing and unboxing types.

rr-codes commented on Source/WebGPU/WebGPU/UsdModelRenderer.swift:
```diff
@@ -0,0 +1,439 @@
+    var materialIds: [UUID]?
+
+    override init() {
+        self.partCount = 0
```
this init shouldnt exist; the initial values should just be set when creating the property

rr-codes commented on Source/WebGPU/WebGPU/UsdModelRenderer.swift:
```diff
@@ -0,0 +1,439 @@
+}
+
+class Helper {
+    static fileprivate func isNonZero(value: Float) -> Bool {
```
these should all be extension functions on the float/simd types

mwyrzykowski replied:
I could not get this to compile, is there another way to write it?

Note `simd_float4` is a typedef for a `float` with a 16 byte alignment, it has no members

<img width="1035" height="135" alt="Screenshot 2025-09-23 at 9 25 21 PM" src="https://github.com/user-attachments/assets/4001ef5b-ea75-4d97-bdf0-a7b4e8712855" />

mwyrzykowski replied:
`simd_float4` is not a class:
```
typedef __attribute__((__ext_vector_type__(4))) float simd_float4;
```

rr-codes commented on Source/WebGPU/WebGPU/UsdModelRenderer.swift:
```diff
@@ -0,0 +1,439 @@
+
+        #if canImport(DirectDrawBackend)
+        context = DirectDrawBackend.Context.createContext(device: device)
+        scene = self.context?.makeScene()
```
no `self`

rr-codes commented on Source/WebGPU/WebGPU/UsdModelRenderer.swift:
```diff
@@ -0,0 +1,439 @@
+
+    @objc(renderWithTexture:)
+    func render(with texture: MTLTexture) {
+        let captureDescriptor = MTLCaptureDescriptor(from: self.device)
```
no self

rr-codes commented on Source/WebGPU/WebGPU/UsdModelRenderer.swift:
```diff
@@ -0,0 +1,439 @@
+        }
+        guard let camera else {
+            return
+        }
```
guard let scene, let camera else {
    return
}

rr-codes commented on Source/WebGPU/WebGPU/UsdModelRenderer.swift:
```diff
@@ -0,0 +1,439 @@
+            if let materialIds = request.materialIds {
+                try mesh.materials = materialIds.map { materialId in
+                    guard let m = materials[materialId] else {
+                        throw NSError()
```
throwing errors should not be used for control flow; this should probably just be a compactMap

rr-codes commented on Source/WebCore/html/canvas/UsdModelLoader.swift:
```diff
@@ -0,0 +1,678 @@
+internal import os
+
+class USDModelLoader: DDRenderDelegate {
+    // let sender: DirectDrawSender
```
remove comment

rr-codes commented on Source/WebCore/html/canvas/UsdModelLoader.swift:
```diff
@@ -0,0 +1,678 @@
+
+class USDModelLoader: DDRenderDelegate {
+    // let sender: DirectDrawSender
+    let renderer: UsdDDRenderer
```
can any of these be private?

rr-codes commented on Source/WebCore/html/canvas/UsdModelLoader.swift:
```diff
@@ -0,0 +1,678 @@
+    let objcLoader: WebUSDModelLoader
+
+    @Observable
+    class Host: Identifiable {
```
final

rr-codes commented on Source/WebCore/html/canvas/UsdModelLoader.swift:
```diff
@@ -0,0 +1,678 @@
+        let id: UUID
+        var state: State
+
+        enum State {
```
nested types should be before properties

rr-codes commented on Source/WebCore/html/canvas/UsdModelLoader.swift:
```diff
@@ -0,0 +1,678 @@
+    }
+
+    func removeMesh(id: UUID) {
+        if let mesh = self.meshes.first(where: { $0.id == id }),
```
meshes should probably be a dictionary for efficiency

mwyrzykowski replied:
size is usually 1 but can be > 1 so I think dictionary is less efficient

rr-codes commented on Source/WebCore/html/canvas/UsdModelLoader.swift:
```diff
@@ -0,0 +1,678 @@
+
+    fileprivate func updateMesh(webRequest: WebUpdateMeshRequest) {
+        Helper.logger.info("update! modelUpdated")
+        if let mu = modelUpdated {
```
`if let modelUpdated`

rr-codes commented on Source/WebCore/html/canvas/UsdModelLoader.swift:
```diff
@@ -0,0 +1,678 @@
+
+    fileprivate func addMesh(
+        webRequest: WebAddMeshRequest,
+        callback: @escaping (Int32) -> Void = { result in
```
this is a bit bizarre; i think just having two separate overloads would be better and much more readable

mwyrzykowski replied:
This allows addMesh to be asynchronous.

I don't think two separate methods allows this.

rr-codes commented on Source/WebCore/html/canvas/UsdModelLoader.swift:
```diff
@@ -0,0 +1,678 @@
+    fileprivate func addMesh(
+        webRequest: WebAddMeshRequest,
+        callback: @escaping (Int32) -> Void = { result in
+            if 0 == result {
```
result == 0

rr-codes commented on Source/WebCore/html/canvas/UsdModelLoader.swift:
```diff
@@ -0,0 +1,678 @@
+    ) {
+        modelAddedCallback = callback
+
+        if let ma = modelAdded {
```
if let modelAdded

achan00 commented on Source/WebCore/Modules/Model/Implementation/DDMeshImpl.cpp:
```diff
@@ -0,0 +1,103 @@
+        .indices = WTFMove(descriptor.indices),
+        .transform = WTFMove(descriptor.transform),
+        .instanceTransforms4x4 = descriptor.instanceTransforms4x4[0].simdAddress(),
+        .instanceTransforms4x4Count = static_cast<uint32_t>(descriptor.instanceTransforms4x4.size()),
```
Storing a raw pointer to an array along with its count is fragile - is it possible for `instanceTransforms4x4` to be a proper `Vector`? Then we don't need the count either.

achan00 commented on Source/WebCore/Modules/Model/InternalAPI/DDFloat4x4.h:
```diff
@@ -0,0 +1,70 @@
+    }
+
+    simd_float4x4* simdAddress() { return &v; }
+    const simd_float4x4* simdAddress() const { return &v; }
```
If we move to the vector, then we can remove `simdAddress()` too?

mwyrzykowski replied:
Yes, will remove it, it's unused now

achan00 commented on Source/WebCore/Modules/Model/InternalAPI/DDMesh.h:
```diff
@@ -0,0 +1,66 @@
+    String m_label;
+};
+
+} // namespace WebCore::WebGPU
```
Nit: `// namespace WebCore::DDModel`
(Ditto for the other files as well)

mwyrzykowski replied:
removing the other half dozen instances

achan00 commented on Source/WebCore/Modules/Model/ModelDDInlineConverters.h:
```diff
@@ -0,0 +1,167 @@
+static Vector<KeyValuePair<int32_t, uint64_t>> toCpp(NSArray<WebSetRenderFlags *> *renderFlags)
+{
+    Vector<KeyValuePair<int32_t, uint64_t>> result;
+    for (WebSetRenderFlags* flag in renderFlags)
```
Nit: `WebSetRenderFlags *flag in...`

achan00 commented on Source/WebCore/html/canvas/UsdModelLoader.swift:
```diff
@@ -0,0 +1,713 @@
+        indices: Data?,
+        transform: simd_float4x4,
+        materialIds: [UUID]?
+    ) {
```
How come `WebUpdateMeshRequest`'s init doesn't take in `instanceTransforms`?

mwyrzykowski replied:
The Swift type `[simd_float4x4]` from the underlying framework can not be represented in Objective-C because Objective-C does not allow value types in NSArray

mwyrzykowski replied:
I applied your suggestion below, that worked to add it to the initializer :)

achan00 commented on Source/WebCore/html/canvas/UsdModelLoader.swift:
```diff
@@ -0,0 +1,713 @@
+                instanceTransforms?.next = WebChainedFloat4x4(transform: request.instanceTransforms[i])
+                instanceTransforms = instanceTransforms?.next
+            }
+        }
```
Could we compute the instance transforms array and then pass it to the `WebUpdateMeshRequest` initializer?

mwyrzykowski replied:
Yes that would work

achan00 commented on Source/WebGPU/WebGPU/Instance.mm:
```diff
@@ -37,10 +38,26 @@
+        [devices addObject:device];
+#endif
+    return devices;
+}
```
Is this called often? Just wondering if it should be cached with invalidation or it's not worth the extra complexity since it's not called often.

mwyrzykowski replied:
it's not called often but we probably don't need to cache it anyways since it just copies an NSArray of size 1 internally. 

`id<MTLDevice>` is a process wide singleton

achan00 commented on Source/WebGPU/WebGPU/Instance.mm:
```diff
@@ -52,6 +69,11 @@
+    m_ddReceiver = [[DDBridgeReceiver alloc] init];
+    [m_ddReceiver setDeviceWithDevice:getDevices()[0]];
+#endif
+    m_ddMeshIdentifier = [[NSUUID alloc] initWithUUIDString:@"E621E1F8-C36C-495A-93FC-0C247A3E6E5F"];
```
Do we mean to hardcode this UUID?

mwyrzykowski replied:
No, thank you for noticing, I should just call `init` here

achan00 commented on Source/WebGPU/WebGPU/UsdModelRenderer.swift:
```diff
@@ -0,0 +1,491 @@
+        do {
+            try captureManager?.startCapture(with: captureDescriptor)
+        } catch {
+            Logger.modelGPU.info("failed to start gpu capture")
```
Nit: should be an error log

achan00 commented on Source/WebGPU/WebGPU/UsdModelRenderer.swift:
```diff
@@ -0,0 +1,491 @@
+            return
+        }
+
+        self.context?.render(scene: scene, camera: camera)
```
In case something goes wrong, it may be clearer what went wrong if we unwrap `context`, `scene`, `camera` in the beginning and log an error if any is missing. Then we can avoid optional chaining afterwards.

achan00 commented on Source/WebGPU/WebGPU/UsdModelRenderer.swift:
```diff
@@ -0,0 +1,491 @@
+            scene.meshes.append(meshInstance)
+            return true
+        } catch {
+            Logger.modelGPU.info("unexpected error")
```
Consider using error log and print some info about the error.

achan00 commented on Source/WebGPU/WebGPU/UsdModelRenderer.swift:
```diff
@@ -0,0 +1,491 @@
+                    guard let m = materials[materialId] else {
+                        return nil
+                    }
+                    return m
```
Could this just return `materials[materialId]` directly?

achan00 commented on Source/WebGPU/WebGPU/UsdModelRenderer.swift:
```diff
@@ -0,0 +1,491 @@
+                }
+                if let vertices = request.vertices {
+                    for replaceVertices in vertices {
+                        unsafe resource.resource.replaceVertices(index: replaceVertices.bufferIndex) { destination, destinationSize in
```
Maybe worth adding a comment about why we don't need to check `bufferIndex` here because it's done elsewhere.

mwyrzykowski replied:
Added a comment

achan00 commented on Source/WebGPU/WebGPU/UsdModelRenderer.swift:
```diff
@@ -0,0 +1,491 @@
+                                guard let baseAddress = sourceBuffer.baseAddress else {
+                                    return
+                                }
+                                unsafe destination.copyMemory(from: baseAddress, byteCount: destinationSize)
```
Should `byteCount` be just `sourceSize` since it could smaller than `destinationSize`?

mwyrzykowski replied:
yes! thank you for noticing that

achan00 commented on Source/WebGPU/WebGPU/UsdModelRenderer.swift:
```diff
@@ -0,0 +1,491 @@
+                            guard let baseAddress = sourceBuffer.baseAddress else {
+                                return
+                            }
+                            unsafe destination.copyMemory(from: baseAddress, byteCount: destinationSize)
```
Ditto

achan00 commented on Source/WebGPU/WebGPU/UsdModelRenderer.swift:
```diff
@@ -0,0 +1,491 @@
+            }
+            var instTransforms: [simd_float4x4] = []
+            var chainedTransform = request.instanceTransforms
+            while chainedTransform != nil {
```
I think you can use a `while let ...` so you can unwrap `chainedTransform` at the same time.

mwyrzykowski replied:
but then I can not re-assign it two lines later, i.e.,:
```
chainedTransform = chainedTransform.next
```
is an error

mwyrzykowski replied:
`Cannot assign to value: 'chainedTransform' is a 'let' constant`

mwyrzykowski replied:
I found something:
```
            if let chainedTransform = request.instanceTransforms {
                instTransforms.append(chainedTransform.transform)
                while let chainedTransform = chainedTransform.next {
                    instTransforms.append(chainedTransform.transform)
                }
            }
```
that seems fine :)

achan00 commented on Source/WebGPU/WebGPU/UsdModelRenderer.swift:
```diff
@@ -0,0 +1,491 @@
+            }
+            if !instTransforms.isEmpty {
+                mesh.instanceTransforms = instTransforms
+                Logger.modelGPU.info("\(mesh.instanceTransforms)")
```
Will this print out a lot of things (and possibly hurt perf)?

mwyrzykowski replied:
It was for debugging but I think I should remove it now

achan00 commented on Source/WebGPU/WebGPU/UsdModelRenderer.swift:
```diff
@@ -0,0 +1,491 @@
+                Logger.modelGPU.info("\(mesh.instanceTransforms)")
+            }
+        } catch {
+            Logger.modelGPU.info("there was an error")
```
Maybe good to include what method fails in the error log so can track it if it does show up.

mwyrzykowski committed:
Add support for loading USDs in WCP and rendering them in GPUP
https://bugs.webkit.org/show_bug.cgi?id=298819
rdar://160530763

Reviewed by Ada Chan.

Setup framework for sending geometry, transforms, textures -> everything to render a USD,
from WCP to GPUP.

* Source/WebCore/Configurations/WebCore.xcconfig:
* Source/WebCore/Headers.cmake:
* Source/WebCore/Modules/Model/Implementation/DDMeshImpl.cpp: Added.
(WebCore::DDModel::DDMeshImpl::DDMeshImpl):
(WebCore::DDModel::DDMeshImpl::setLabelInternal):
(WebCore::DDModel::convertToBacking):
(WebCore::DDModel::DDMeshImpl::update):
* Source/WebCore/Modules/Model/Implementation/DDMeshImpl.h: Copied from Source/WebCore/Modules/WebGPU/Implementation/WebGPUCreateImpl.cpp.
* Source/WebCore/Modules/Model/Implementation/ModelConvertToBackingContext.cpp: Added.
* Source/WebCore/Modules/Model/Implementation/ModelConvertToBackingContext.h: Added.
* Source/WebCore/Modules/Model/Implementation/ModelDowncastConvertToBackingContext.cpp: Added.
(WebCore::DDModel::DowncastConvertToBackingContext::convertToBacking):
* Source/WebCore/Modules/Model/Implementation/ModelDowncastConvertToBackingContext.h: Added.
* Source/WebCore/Modules/Model/InternalAPI/DDMesh.h: Added.
(WebCore::DDModel::DDMesh::label const):
(WebCore::DDModel::DDMesh::setLabel):
* Source/WebCore/Modules/Model/InternalAPI/DDMeshDescriptor.h: Added.
* Source/WebCore/Modules/Model/InternalAPI/DDMeshPart.h: Added.
* Source/WebCore/Modules/Model/InternalAPI/DDReplaceVertices.h: Added.
* Source/WebCore/Modules/Model/InternalAPI/DDUpdateMeshDescriptor.h: Added.
* Source/WebCore/Modules/Model/InternalAPI/DDVertexAttributeFormat.h: Added.
* Source/WebCore/Modules/Model/InternalAPI/DDVertexLayout.h: Added.
* Source/WebCore/Modules/Model/InternalAPI/ModelObjectDescriptorBase.h: Added.
* Source/WebCore/Modules/WebGPU/Implementation/WebGPUCreateImpl.cpp:
(WebCore::WebGPU::create):
* Source/WebCore/Modules/WebGPU/Implementation/WebGPUImpl.cpp:
(WebCore::WebGPU::GPUImpl::GPUImpl):
(WebCore::WebGPU::convertToBacking):
(WebCore::WebGPU::GPUImpl::addMeshRequest):
* Source/WebCore/Modules/WebGPU/Implementation/WebGPUImpl.h:
* Source/WebCore/Modules/WebGPU/Implementation/WebGPUPtr.h:
* Source/WebCore/Modules/WebGPU/InternalAPI/WebGPU.h:
* Source/WebCore/Modules/mediasource/ManagedMediaSource.cpp:
* Source/WebCore/Modules/mediastream/RTCRtpSFrameTransformerCocoa.cpp:
* Source/WebCore/Sources.txt:
* Source/WebCore/SourcesCocoa.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/html/canvas/UsdModelLoader.swift: Added.
(USDModelLoader.timer):
(Host.state):
(Connected.hosts):
(state):
(meshes):
(primIdToMeshId):
(materials):
(primIdToMaterialId):
(materialIdToTextureIds):
(assetPathToTextureId):
(loadModel(from:)):
(addHost):
(updateHost(_:size:contentsScale:)):
(removeHost(_:)):
(removeMesh(_:)):
(setCameraDistance(_:)):
(setEnableModelRotation(_:)):
(meshCreated(_:)):
(meshUpdated(_:data:)):
(meshDestroyed(_:)):
(materialCreated(_:)):
(materialUpdated(_:data:)):
(materialDestroyed(_:)):
(textureCreated(_:)):
(textureUpdated(_:data:)):
(textureDestroyed(_:)):
(WebAddMeshRequest.indexCapacity):
(WebAddMeshRequest.indexType):
(WebAddMeshRequest.vertexBufferCount):
(WebAddMeshRequest.vertexCapacity):
(WebAddMeshRequest.vertexAttributes):
(WebAddMeshRequest.vertexLayouts):
(WebDDMeshPart.indexOffset):
(WebDDMeshPart.indexCount):
(WebDDMeshPart.topology):
(WebDDMeshPart.materialIndex):
(WebDDMeshPart.boundsMin):
(WebDDMeshPart.boundsMax):
(WebSetPart.partIndex):
(WebSetPart.part):
(WebSetRenderFlags.partIndex):
(WebSetRenderFlags.renderFlags):
(WebReplaceVertices.bufferIndex):
(WebReplaceVertices.buffer):
(WebUpdateMeshRequest.partCount):
(WebUpdateMeshRequest.parts):
(WebUpdateMeshRequest.renderFlags):
(WebUpdateMeshRequest.vertices):
(WebUpdateMeshRequest.indices):
(WebUpdateMeshRequest.transform):
(WebUpdateMeshRequest.instanceTransforms):
(WebUpdateMeshRequest.materialIds):
(WebUSDModelLoader.renderer):
(WebUSDModelLoader.modelAdded):
(WebUSDModelLoader.modelUpdated):
(WebUSDModelLoader.modelAddedCallback):
(WebUSDModelLoader.setCallbacks(_:modelUpdatedCallback:)):
(WebUSDModelLoader.loadModel(from:)):
(WebUSDModelLoader.toFloatArray(_:)):
(WebUSDModelLoader.results):
(WebUSDModelLoader.toWebSetParts(_:)):
(WebUSDModelLoader.toWebRenderFlags(_:)):
(WebUSDModelLoader.toWebReplaceVertices(_:)):
(WebUSDModelLoader.toWebVertexAttributes(_:)):
(WebUSDModelLoader.toWebVertexLayouts(_:)):
(WebUSDModelLoader.updateMesh(_:)):
(WebUSDModelLoader.addMesh(_:callback:)):
(WebUSDModelLoader.meshAdded(_:)):
(WebUSDModelLoader.updateHostedRenderer(_:size:contentsScale:)):
* Source/WebGPU/WebGPU.xcodeproj/project.pbxproj:
* Source/WebGPU/WebGPU/CommandEncoder.swift:
(WebGPU.beginRenderPass(_:)):
* Source/WebGPU/WebGPU/Instance.h:
* Source/WebGPU/WebGPU/Instance.mm:
(WebGPU::getDevices):
(WebGPU::m_isValid):
(WebGPU::Instance::requestAdapter):
(WebGPU::Instance::updateModel):
(WebGPU::convertPart):
(WebGPU::convertUint8s):
(WebGPU::convertDescriptor):
(WebGPU::Instance::updateMesh):
(wgpuDDMeshCreate):
* Source/WebGPU/WebGPU/Internal/Logging.h:
* Source/WebGPU/WebGPU/UsdModelRenderer.swift: Added.
(DDBridgeAddMeshRequest.indexCapacity):
(DDBridgeAddMeshRequest.indexType):
(DDBridgeAddMeshRequest.vertexBufferCount):
(DDBridgeAddMeshRequest.vertexCapacity):
(DDBridgeAddMeshRequest.vertexAttributes):
(DDBridgeAddMeshRequest.vertexLayouts):
(DDBridgeAddMeshRequest.makeDescriptor):
(DDMeshPart.indexOffset):
(DDMeshPart.indexCount):
(DDMeshPart.topology):
(DDMeshPart.materialIndex):
(DDMeshPart.boundsMin):
(DDMeshPart.boundsMax):
(DDBridgeSetPart.partIndex):
(DDBridgeSetPart.part):
(DDBridgeSetRenderFlags.partIndex):
(DDBridgeSetRenderFlags.renderFlags):
(DDBridgeReplaceVertices.bufferIndex):
(DDBridgeReplaceVertices.buffer):
(DDBridgeUpdateMesh.partCount):
(DDBridgeUpdateMesh.parts):
(DDBridgeUpdateMesh.renderFlags):
(DDBridgeUpdateMesh.vertices):
(DDBridgeUpdateMesh.indices):
(DDBridgeUpdateMesh.transform):
(DDBridgeUpdateMesh.instanceTransforms):
(DDBridgeUpdateMesh.materialIds):
(DDReceiver.device):
(DDReceiver.context):
(DDReceiver.scene):
(DDReceiver.meshes):
(DDReceiver.materials):
(DDReceiver.nodeDefinitionStore):
(DDReceiver.camera):
(DDReceiver.commandQueue):
(DDReceiver.captureManager):
(DDReceiver.setDevice(_:)):
(DDReceiver.makeCaptureDescriptor() -> (MTLCaptureDescriptor:Foundation:)):
(DDReceiver.render(_:)):
(DDReceiver.addMesh(_:identifier:)):
(DDReceiver.updateMesh(_:identifier:)):
* Source/WebGPU/WebGPU/WebGPUExt.h:
* Source/WebKit/CMakeLists.txt:
* Source/WebKit/DerivedSources-input.xcfilelist:
* Source/WebKit/DerivedSources-output.xcfilelist:
* Source/WebKit/DerivedSources.make:
* Source/WebKit/GPUProcess/graphics/Model/ModelObjectHeap.cpp: Added.
(WebKit::DDModel::ObjectHeap::ObjectHeap):
(WebKit::DDModel::ObjectHeap::addObject):
(WebKit::DDModel::ObjectHeap::removeObject):
(WebKit::DDModel::ObjectHeap::clear):
(WebKit::DDModel::ObjectHeap::convertDDMeshFromBacking):
(WebKit::DDModel::ObjectHeap::objectExistsAndValid const):
* Source/WebKit/GPUProcess/graphics/Model/ModelObjectHeap.h: Added.
* Source/WebKit/GPUProcess/graphics/Model/RemoteDDMesh.cpp: Copied from Source/WebKit/GPUProcess/graphics/WebGPU/RemoteBindGroup.cpp.
(WebKit::RemoteDDMesh::RemoteDDMesh):
(WebKit::RemoteDDMesh::connection const):
(WebKit::RemoteDDMesh::stopListeningForIPC):
(WebKit::RemoteDDMesh::destruct):
(WebKit::RemoteDDMesh::setLabel):
(WebKit::RemoteDDMesh::protectedBacking):
(WebKit::RemoteDDMesh::update):
* Source/WebKit/GPUProcess/graphics/Model/RemoteDDMesh.h: Added.
* Source/WebKit/GPUProcess/graphics/Model/RemoteDDMesh.messages.in: Copied from Source/WebKit/GPUProcess/graphics/WebGPU/RemoteGPU.messages.in.
* Source/WebKit/GPUProcess/graphics/WebGPU/RemoteBindGroup.cpp:
* Source/WebKit/GPUProcess/graphics/WebGPU/RemoteGPU.cpp:
(WebKit::RemoteGPU::RemoteGPU):
(WebKit::RemoteGPU::workQueueUninitialize):
(WebKit::RemoteGPU::addMeshRequest):
* Source/WebKit/GPUProcess/graphics/WebGPU/RemoteGPU.h:
* Source/WebKit/GPUProcess/graphics/WebGPU/RemoteGPU.messages.in:
* Source/WebKit/GPUProcess/webrtc/RemoteAudioMediaStreamTrackRendererInternalUnitManager.cpp:
(WebKit::RemoteAudioMediaStreamTrackRendererInternalUnitManagerUnit::RemoteAudioMediaStreamTrackRendererInternalUnitManagerUnit):
* Source/WebKit/Scripts/webkit/messages.py:
(atomic_object_identifier):
(serialized_identifiers):
(headers_for_type):
* Source/WebKit/Shared/Model/DDMeshDescriptor.cpp: Added.
(WebKit::DDModel::ConvertToBackingContext::convertToBacking):
(WebKit::DDModel::convertVectorToBacking):
(WebKit::DDModel::ConvertFromBackingContext::convertFromBacking):
(WebKit::DDModel::convertVectorFromBacking):
* Source/WebKit/Shared/Model/DDMeshDescriptor.h: Added.
* Source/WebKit/Shared/Model/DDMeshDescriptor.serialization.in: Added.
* Source/WebKit/Shared/Model/DDMeshPart.cpp: Added.
(WebKit::DDModel::ConvertToBackingContext::convertToBacking):
(WebKit::DDModel::ConvertFromBackingContext::convertFromBacking):
* Source/WebKit/Shared/Model/DDMeshPart.h: Added.
* Source/WebKit/Shared/Model/DDMeshPart.serialization.in: Added.
* Source/WebKit/Shared/Model/DDModelIdentifier.h: Added.
* Source/WebKit/Shared/Model/DDReplaceVertices.cpp: Added.
(WebKit::DDModel::ConvertToBackingContext::convertToBacking):
(WebKit::DDModel::ConvertFromBackingContext::convertFromBacking):
* Source/WebKit/Shared/Model/DDReplaceVertices.h: Added.
* Source/WebKit/Shared/Model/DDReplaceVertices.serialization.in: Added.
* Source/WebKit/Shared/Model/DDUpdateMeshDescriptor.cpp: Added.
(WebKit::DDModel::ConvertToBackingContext::convertToBacking):
(WebKit::DDModel::convertVectorToBacking):
(WebKit::DDModel::ConvertFromBackingContext::convertFromBacking):
(WebKit::DDModel::convertVectorFromBacking):
* Source/WebKit/Shared/Model/DDUpdateMeshDescriptor.h: Added.
* Source/WebKit/Shared/Model/DDUpdateMeshDescriptor.serialization.in: Added.
* Source/WebKit/Shared/Model/DDVertexAttributeFormat.h: Added.
* Source/WebKit/Shared/Model/DDVertexAttributeFormat.serialization.in: Added.
* Source/WebKit/Shared/Model/DDVertexLayout.h: Added.
* Source/WebKit/Shared/Model/DDVertexLayout.serialization.in: Added.
* Source/WebKit/Shared/Model/ModelConvertFromBackingContext.h: Added.
* Source/WebKit/Shared/Model/ModelConvertToBackingContext.h: Added.
* Source/WebKit/Shared/Model/ModelObjectDescriptorBase.cpp: Added.
(WebKit::DDModel::ConvertToBackingContext::convertToBacking):
(WebKit::DDModel::ConvertFromBackingContext::convertFromBacking):
* Source/WebKit/Shared/Model/ModelObjectDescriptorBase.h: Added.
* Source/WebKit/Shared/Model/ModelObjectDescriptorBase.serialization.in: Added.
* Source/WebKit/Shared/WTFArgumentCoders.serialization.in:
* Source/WebKit/Sources.txt:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/GPU/graphics/Model/ModelDowncastConvertToBackingContext.cpp: Added.
(WebKit::DDModel::DowncastConvertToBackingContext::convertToBacking):
* Source/WebKit/WebProcess/GPU/graphics/Model/ModelDowncastConvertToBackingContext.h: Added.
* Source/WebKit/WebProcess/GPU/graphics/Model/RemoteDDMeshProxy.cpp: Added.
(WebKit::DDModel::RemoteDDMeshProxy::RemoteDDMeshProxy):
(WebKit::DDModel::RemoteDDMeshProxy::~RemoteDDMeshProxy):
(WebKit::DDModel::RemoteDDMeshProxy::update):
(WebKit::DDModel::RemoteDDMeshProxy::setLabelInternal):
* Source/WebKit/WebProcess/GPU/graphics/Model/RemoteDDMeshProxy.h: Added.
* Source/WebKit/WebProcess/GPU/graphics/WebGPU/RemoteGPUProxy.cpp:
(WebKit::RemoteGPUProxy::create):
(WebKit::RemoteGPUProxy::RemoteGPUProxy):
(WebKit::RemoteGPUProxy::addMeshRequest):
* Source/WebKit/WebProcess/GPU/graphics/WebGPU/RemoteGPUProxy.h:
* Source/WebKit/WebProcess/GPU/media/RemoteAudioDestinationProxy.cpp:
(WebKit::RemoteAudioDestinationProxy::outputLatency const):
(WebKit::RemoteAudioDestinationProxy::renderAudio):
* Source/WebKit/WebProcess/Network/webrtc/LibWebRTCNetworkManager.cpp:
* Source/WebKit/WebProcess/Storage/WebSWContextManagerConnection.cpp:
* Source/WebKit/WebProcess/WebCoreSupport/WebChromeClient.cpp:
(WebKit::WebChromeClient::createGPUForWebGPU const):
* Source/WebKit/WebProcess/WebCoreSupport/WebPermissionController.cpp:
(WebKit::WebPermissionController::query):
* Source/WebKit/WebProcess/WebCoreSupport/WebWorkerClient.cpp:
(WebKit::GPUProcessWebWorkerClient::createGPUForWebGPU const):

Canonical link: https://commits.webkit.org/300626@main
